File: frontend/src/types/panel/services.ts
/** Mirrors app/models/panel/services.py */
import type { PagedRequest, PageMeta, PanelAuthRequest, PanelEnvelope, PanelOption } from "./common";

export interface PanelServiceRow {
  code: number;
  user_id?: number | null;
  username?: string | null;
  panel?: string | null;
  panel_code?: number | null;
  package_size?: number | null;
  expiration_time?: number | null;
  enable: boolean;
  expired: boolean;
  is_test: boolean;
}

export interface PanelServicesRequest extends PagedRequest {
  q?: string;
  /** Panel code, or empty for all. */
  panel?: string;
  /** empty | active | expired | test */
  state?: string;
}

export interface PanelServicesResponse extends PanelEnvelope {
  services: PanelServiceRow[];
  panels: PanelOption[];
  meta: PageMeta;
}

export interface PanelServiceToggleRequest extends PanelAuthRequest {
  code: number;
  enabled: boolean;
}

export interface PanelServiceDeleteRequest extends PanelAuthRequest {
  code: number;
}

export interface PanelTransactionRow {
  id: number;
  user_id?: number | null;
  amount: number;
  status?: string | null;
  method?: string | null;
  created_at?: number | null;
  receipt?: string | null;
}

export interface PanelTransactionsRequest extends PagedRequest {
  /** empty | pending | approved | rejected */
  status?: string;
  /** empty | card | crypto */
  method?: string;
}

export interface PanelTransactionsResponse extends PanelEnvelope {
  transactions: PanelTransactionRow[];
  meta: PageMeta;
  pending_total: number;
}

export interface PanelTransactionActionRequest extends PanelAuthRequest {
  tx_id: number;
}

File: app/models/panel/services.py
"""Admin panel DTOs: sold services and payment transactions."""

from pydantic import BaseModel, Field

from app.models.panel.common import PagedRequest, PageMeta, PanelRequest, PanelResponse


class PanelServiceRow(BaseModel):
    code: int
    user_id: int | None = None
    username: str | None = None
    panel: str | None = None
    panel_code: int | None = None
    package_size: float | None = None
    expiration_time: int | None = None
    enable: bool = True
    expired: bool = False
    is_test: bool = False


class PanelServicesRequest(PagedRequest):
    q: str = Field("", max_length=64)
    panel: str = Field("", description="Panel code, or empty for all")
    state: str = Field("", description="empty | active | expired | test")


class PanelPanelOption(BaseModel):
    code: int
    name: str


class PanelServicesResponse(PanelResponse):
    services: list[PanelServiceRow] = Field(default_factory=list)
    panels: list[PanelPanelOption] = Field(default_factory=list)
    meta: PageMeta = Field(default_factory=PageMeta)


class PanelServiceToggleRequest(PanelRequest):
    code: int
    enabled: bool


class PanelServiceDeleteRequest(PanelRequest):
    code: int


class PanelTransactionRow(BaseModel):
    id: int
    user_id: int | None = None
    amount: int = 0
    status: str | None = None
    method: str | None = None
    created_at: int | None = None
    receipt: str | None = None


class PanelTransactionsRequest(PagedRequest):
    status: str = Field("", description="empty | pending | approved | rejected")
    method: str = Field("", description="empty | card | crypto")


class PanelTransactionsResponse(PanelResponse):
    transactions: list[PanelTransactionRow] = Field(default_factory=list)
    meta: PageMeta = Field(default_factory=PageMeta)
    pending_total: int = 0


class PanelTransactionActionRequest(PanelRequest):
    tx_id: int
